---
title: Lostruct
date: 2021-02-12
author: Kevin Murray
---

```{r setup, include=F, cache=F}

if (!require("tidyverse"))         { install.packages("tidyverse")                              ; require("tidyverse")          }
if (!require("foreach"))           { install.packages("foreach")                                ; require("foreach")            }
if (!require("doParallel"))        { install.packages("doParallel")                             ; require("doParallel")         }
if (!require("parallel"))          { install.packages("parallel")                               ; require("parallel")           }
if (!require("ggplot2"))           { install.packages("ggplot2")                                ; require("ggplot2")            }
if (!require("GGally"))            { install.packages("GGally")                                 ; require("GGally")             }
if (!require("ggrepel"))           { install.packages("ggrepel")                                ; require("ggrepel")            }
if (!require("ggmap"))             { install.packages("ggmap")                                  ; require("ggmap")              }
if (!require("RColorBrewer"))      { install.packages("RColorBrewer")                           ; require("RColorBrewer")       }
if (!require("SNPRelate"))         { BiocManager::install("SNPRelate")                          ; require("SNPRelate")          }
if (!require("lostruct"))          { remotes::install_github("petrelharp/local_pca/lostruct")   ; require("lostruct")           }
if (!require("windowlickr"))       { remotes::install_github("kdm9/windowlickr")                ; require("windowlickr")        }

knitr::opts_chunk$set(
  warnings=F
)
theme_set(theme_bw())

NCPUS = as.integer(Sys.getenv("NCPUS", parallel::detectCores(logical=F)))

```

## BCF windower

So at time of writing, lostruct's vcf parsing code is borked. So, I've written a custom bit of code using windowlickr to do the snp extraction on windows.

```{r}

windowlickr_windowfun <- function (file, windows=NULL, size=20000, samples=NULL, ...) {
    if (is.null(samples)) { samples = windowlickr:::bcf_getSamples(file) }
    if (is.null(windows)) { windows = windowlickr:::bcf_getWindows(file, windowsize=size, slide=size)}
    pos.fn <- function(n, ...) {
        windows[n,] %>%
            dplyr::select(chrom=contig, start, end=stop)
    }
    win.fn <- function (n,...) {
        if (n > nrow(windows)) stop(paste0("No such window: ", n))
        region = windows[n, "region"]
        ret = windowlickr:::readBCFQuery_(file, region, samples)
        nsnp = length(ret$POS)
        if (nsnp < 10) {
            return(NULL)
        }
        GT = matrix(unlist(ret$GT, recursive = F), nrow=nsnp, byrow = T)
        return(GT)
    }
    attr(win.fn,"max.n") <- nrow(windows)
    attr(win.fn,"region") <- pos.fn
    attr(win.fn,"samples") <- samples
    class(win.fn) <- c("winfun", "function")
    return(win.fn)
}

```

### Data prep for eigen_windows


```{r}

# params: window size & input file
winsize=200000
file="data/variants/AdnatariaHighCoverage_Emelref.norepeat.vcf.gz"
npc=20
load("data/cache/02_filterthresholds.Rda")

# Get genome windows and windower function for eigen_windows
windows = windowlickr:::bcf_getWindows(file, windowsize=winsize, slide=winsize) %>%
    filter(!grepl("^tig", contig))
table(windows$contig)
wf = windowlickr_windowfun(file=file, windows=windows, size=winsize, samples=good.samp.em)
```

### Heavy calculations: eigenvectors per window, then distances among windows

```{r}
# Calculate pop. struct. (eigenvectors) of each window
eigwin = xfun:::cache_rds({
    eigen_windows(wf, k=npc, mc.cores=NCPUS)
}, dir="data/cache/", file="04_lostruct-Em-eigenwindows.Rds")

# calculate inter-window distances
pcd = xfun::cache_rds({
    pc_dist(eigwin, npc=npc)
}, dir="data/cache/", file="04_lostruct-Em-pcdistances.Rds")
```


### Plotting

```{r}
# exclude NA windows
na.window.idx = is.na(eigwin[,1])
table(na.window.idx)
non.na.windows = windows[!na.window.idx,]

# mds
mds = cmdscale( pcd[!na.window.idx,!na.window.idx], k=5, eig=T)

# make data frame
pts = mds$points
colnames(pts) = paste0("MDS",seq_len(ncol(pts)))
mds.dat = bind_cols(non.na.windows, as.data.frame(pts))
str(mds.dat)
ggpairs(as.data.frame(pts), upper="blank")
ggsave("plot/04_lostruct-Em-mdspairs.svg")
```

```{r plot.mds}
ctgs = windowlickr:::bcf_getContigs(file)
ctg.lens=cumsum(c(0, ctgs$lengths[-length(ctgs$lengths)]))
names(ctg.lens) = ctgs$names

mds.plt = mds.dat %>%
    gather("mds.dim", "value", starts_with("MDS")) %>%
    mutate(overall_pos = ctg.lens[contig] + start)


colours = rep(c("#1f78b4", "#a6cee3"), length.out=nrow(ctgs))
shapes = rep(c(16, 15), length.out=nrow(ctgs))
ggplot(mds.plt, aes(x=overall_pos, y=value)) +
    geom_point(aes(colour=contig, shape=contig)) +
    facet_grid(mds.dim~., space="free_x") +
    scale_colour_manual(values=colours) +
    scale_shape_manual(values=shapes) +
    labs(x="Genome position", y="MDS axis") +
    theme(legend.position="none", panel.grid=element_blank())
ggsave("plot/04_lostruct-Em-manhattan.svg")

```
